src/eos-refcounted.h

```diff
@@ -0,0 +1,74 @@
+  G_DEFINE_TYPE (TypeName, type_name, G_TYPE_OBJECT)                    \
+                                                                        \
+  static void                                                           \
+  type_name##_real_dispose (GObject *object)                            \
+  {                                                                     \
+    typedef void (*Func)(TypeName *obj);                                \
+                                                                        \
+    TypeName *self = TYPE_NAME (object);                                \
+    Func f = dispose_impl_func;                                         \
+                                                                        \
+    if (f != NULL)                                                      \
+      f (self);                                                         \
+    G_OBJECT_CLASS (type_name##_parent_class)->dispose (object);        \
+  }                                                                     \
+                                                                        \
+  static void                                                           \
+  type_name##_real_finalize (GObject *object)                           \
+  {                                                                     \
+    typedef void (*Func)(TypeName *obj);                                \
+                                                                        \
+    TypeName *self = TYPE_NAME (object);                                \
+    Func f = finalize_impl_func;                                        \
+                                                                        \
+    if (f != NULL)                                                      \
+      f (self);                                                         \
+    G_OBJECT_CLASS (type_name##_parent_class)->finalize (object);       \
+  }                                                                     \
+                                                                        \
+  static void                                                           \
+  type_name##_class_init (TypeName##Class *self_class)                  \
+  {                                                                     \
+    GObjectClass *object_class = G_OBJECT_CLASS (self_class);           \
+                                                                        \
+    object_class->dispose = type_name##_real_dispose;                   \
+    object_class->finalize = type_name##_real_finalize;                 \
+  }                                                                     \
+                                                                        \
+  static void                                                           \
+  type_name##_init (TypeName *self)                                     \
+  {}
+
+G_END_DECLS
```

Comment:
Hm, okay, it seems to me like this wraps things which are already GObjects but might not necessarily have a dispose/finalise func?